File: internal/app/saves.go
package app

import (
	"context"
	"errors"
	"fmt"
	"os/exec"
	"strings"
	"sync"
	"time"

	"github.com/ApolloF/WaterLauncher/internal/launch"
	"github.com/ApolloF/WaterLauncher/internal/library"
	"github.com/ApolloF/WaterLauncher/internal/logx"
	"github.com/ApolloF/WaterLauncher/internal/platform"
	"github.com/ApolloF/WaterLauncher/internal/syncer"
)

// syncerPage is where Syncer can be downloaded.
const syncerPage = "https://github.com/ApolloF/syncer/releases/latest"

// Saves is what Syncer knows about a game's saves, for the interface.
type Saves struct {
	Installed bool            `json:"installed"` // Syncer is installed
	Outdated  bool            `json:"outdated"`  // too old for the launcher API
	Available bool            `json:"available"` // Syncer answered
	Error     string          `json:"error,omitempty"`
	Known     bool            `json:"known"` // Syncer has save folders for this game
	Folders   []syncer.Folder `json:"folders"`
}

// SyncerStatus is how WaterLauncher and Syncer get on, for Settings.
type SyncerStatus struct {
	Installed   bool   `json:"installed"`
	Version     string `json:"version,omitempty"`
	Outdated    bool   `json:"outdated"`  // too old for the launcher API
	Connected   bool   `json:"connected"` // Syncer answered
	Running     bool   `json:"running"`   // its launcher API is up (it may not have been asked to start)
	Error       string `json:"error,omitempty"`
	Syncing     bool   `json:"syncing"` // its sync engine runs
	Paused      bool   `json:"paused"`
	PausedUntil int64  `json:"pausedUntil,omitempty"`
	BackingUp   bool   `json:"backingUp"`
	LastBackup  int64  `json:"lastBackup,omitempty"`
	Games       int    `json:"games"`
	Conflicts   int    `json:"conflicts"`
	CheckedAt   int64  `json:"checkedAt"`
}

// Syncer reports Syncer's state. With start it starts Syncer (without its
// window) when it isn't running, as a launch would; otherwise it only
// looks, so opening Settings doesn't start anything.
func (s *SavesService) Syncer(start bool) SyncerStatus {
	out := SyncerStatus{CheckedAt: time.Now().Unix()}
	inst, ok := syncer.Find()
	if !ok {
		return out
	}
	out.Installed, out.Version = true, inst.Version
	if inst.Version != "" && !inst.API() {
		out.Outdated, out.Error = true, syncer.ErrOutdated.Error()
		return out
	}
	ctx, cancel := context.WithTimeout(s.c.ctx, 15*time.Second)
	defer cancel()
	cl, err := syncer.Dial(ctx, start)
	switch {
	case errors.Is(err, syncer.ErrOutdated):
		out.Outdated, out.Error = true, err.Error()
		return out
	case err != nil && !start:
		return out // not running, and not asked to start it
	case err != nil:
		out.Error = "Syncer didn't answer"
		return out
	}
	defer cl.Close()
	out.Running = true
	st, err := cl.Status(ctx)
	if err != nil {
		out.Error = err.Error()
		out.Outdated = strings.Contains(err.Error(), "API version")
		return out
	}
	out.Connected = true
	if st.Version != "" {
		out.Version = strings.TrimPrefix(st.Version, "v")
	}
	out.Syncing, out.Paused, out.BackingUp = st.Syncthing, st.Paused, st.BackingUp
	out.Games, out.Conflicts = st.Games, st.Conflicts
	if !st.PausedTill.IsZero() {
		out.PausedUntil = st.PausedTill.Unix()
	}
	if !st.LastBackup.IsZero() {
		out.LastBackup = st.LastBackup.Unix()
	}
	return out
}

// SavesService is the game's saves, through Syncer.
type SavesService struct {
	c *Core

	mu    sync.Mutex
	cache map[int64]cachedSaves
}

type cachedSaves struct {
	at time.Time
	s  Saves
}

// NewSavesService binds Syncer to core.
func NewSavesService(c *Core) *SavesService {
	return &SavesService{c: c, cache: map[int64]cachedSaves{}}
}

// Saves returns what Syncer knows about a game's saves. Syncer is started
// in the background (without its window) when needed. Answers are kept
// for a little while; fresh asks again right away.
func (s *SavesService) Saves(id int64, fresh bool) (Saves, error) {
	g, ok := s.c.Lib.Get(id)
	if !ok {
		return Saves{}, library.ErrNotFound
	}
	s.mu.Lock()
	if c, ok := s.cache[id]; ok && !fresh && time.Since(c.at) < 30*time.Second {
		s.mu.Unlock()
		return c.s, nil
	}
	s.mu.Unlock()

	out := Saves{Folders: []syncer.Folder{}}
	if _, ok := syncer.Installed(); !ok {
		return out, nil
	}
	out.Installed = true
	ctx, cancel := context.WithTimeout(s.c.ctx, 15*time.Second)
	defer cancel()
	start := s.c.Settings.Get().StartSyncer
	cl, err := syncer.Dial(ctx, start)
	if errors.Is(err, syncer.ErrOutdated) {
		out.Outdated, out.Error = true, err.Error()
		return out, nil
	}
	if err != nil {
		out.Error = "Syncer didn't answer"
		if !start {
			out.Error = "Syncer isn't running"
		}
		return out, nil
	}
	defer cl.Close()
	if _, err := cl.Status(ctx); err != nil {
		out.Error = err.Error()
		out.Outdated = strings.Contains(err.Error(), "API version")
		return out, nil
	}
	out.Available = true
	gs, err := cl.GameStatus(ctx, syncerGame(g))
	if err != nil {
		out.Error = err.Error()
		return out, nil
	}
	out.Known = gs.Known
	if gs.Folders != nil {
		out.Folders = gs.Folders
	}
	s.mu.Lock()
	s.cache[id] = cachedSaves{time.Now(), out}
	s.mu.Unlock()
	return out, nil
}

// OpenSyncer shows Syncer's window.
func (s *SavesService) OpenSyncer() error {
	exe, ok := syncer.Installed()
	if !ok {
		return syncer.ErrNotInstalled
	}
	cmd := exec.Command(exe)
	if err := cmd.Start(); err != nil { // a second start brings the open window forward
		return err
	}
	return cmd.Process.Release()
}

// GetSyncer opens Syncer's download page.
func (s *SavesService) GetSyncer() error { return platform.OpenWebPage(syncerPage) }

// syncerGame describes a game to Syncer.
func syncerGame(g library.Game) syncer.Game {
	app := g.SteamAppID
	if app == 0 {
		app = g.MetaAppID
	}
	return syncer.Game{Title: g.DisplayTitle(), Dir: g.Dir, SteamAppID: app, GogID: g.GogID}
}

// syncerGames is every installed game, for Syncer's matching.
func (c *Core) syncerGames() []syncer.Game {
	var out []syncer.Game
	for _, g := range c.Lib.Games() {
		if g.Installed && !g.Hidden {
			out = append(out, syncerGame(g))
		}
	}
	return out
}

// registerWithSyncer tells a running Syncer which games this PC has. It
// doesn't start Syncer for it.
func (c *Core) registerWithSyncer() {
	if _, ok := syncer.Installed(); !ok {
		return
	}
	ctx, cancel := context.WithTimeout(c.ctx, 10*time.Second)
	defer cancel()
	cl, err := syncer.Dial(ctx, false)
	if err != nil {
		return
	}
	defer cl.Close()
	if err := cl.RegisterGames(ctx, c.syncerGames()); err != nil {
		logx.Printf("syncer: registering games: %v", err)
	}
}

// savesBeforeStep brings the game's saves up to date before it starts:
// it stops for two versions of a save or a newer save still on its way
// from another PC, then waits for Syncer to finish syncing.
func (c *Core) savesBeforeStep(g library.Game, known *bool) launch.Step {
	title := g.DisplayTitle()
	return launch.Step{ID: "savesBefore", Label: "Sync saves", Timeout: 7 * time.Minute,
		Run: func(ctx context.Context, sc *launch.StepContext) error {
			sc.Progress("Asking Syncer…")
			cfg := c.Settings.Get()
			dctx, cancel := context.WithTimeout(ctx, 15*time.Second)
			cl, err := syncer.Dial(dctx, cfg.StartSyncer)
			cancel()
			if errors.Is(err, syncer.ErrOutdated) {
				return err
			}
			if err != nil && !cfg.StartSyncer {
				sc.Progress("Syncer isn't running")
				return nil
			}
			if err != nil {
				return errors.New("Syncer didn't answer")
			}
			defer cl.Close()
			if _, err := cl.Status(ctx); err != nil {
				return err
			}
			_ = cl.RegisterGames(ctx, c.syncerGames())
			gs, err := cl.GameStatus(ctx, syncerGame(g))
			if err != nil {
				return err
			}
			*known = gs.Known
			if !gs.Known {
				sc.Progress("Syncer has no saves for this game")
				return nil
			}
			conflicts, newerOn := 0, ""
			var ids []string
			for _, f := range gs.Folders {
				conflicts += f.Conflicts
				if f.NewerOn != "" && newerOn == "" {
					newerOn = f.NewerOn
				}
				if f.Sync {
					ids = append(ids, f.ID)
				}
			}
			wait := time.Duration(cfg.SyncWait) * time.Second
			if conflicts > 0 {
				a, err := sc.Ask(ctx, fmt.Sprintf("%s has two versions of a save. Pick the one to keep in Syncer first, so you don't play on the wrong one.", title), []launch.Option{
					{ID: "open", Label: "Open Syncer"},
					{ID: "play", Label: "Play anyway"},
					{ID: "cancel", Label: "Cancel"},
				})
				switch {
				case err != nil:
					return err
				case a == "open":
					_ = cl.Open(ctx)
					return launch.ErrCancel
				case a == "cancel":
					return launch.ErrCancel
				}
			}
			if newerOn != "" {
				a, err := sc.Ask(ctx, fmt.Sprintf("%s has a newer save of %s that hasn't arrived on this PC yet.", newerOn, title), []launch.Option{
					{ID: "wait", Label: "Wait for it"},
					{ID: "play", Label: "Play anyway"},
					{ID: "cancel", Label: "Cancel"},
				})
				switch {
				case err != nil:
					return err
				case a == "cancel":
					return launch.ErrCancel
				case a == "wait":
					wait = max(wait, 2*time.Minute+30*time.Second)
				}
			}
			if len(ids) == 0 {
				sc.Progress("Backed up, not synced")
				return nil
			}
			sc.Progress("Syncing…")
			res, err := cl.SyncNow(ctx, ids, wait)
			if err != nil {
				return err
			}
			for _, r := range res {
				if !r.Done {
					if r.State == "paused" {
						return errors.New("Syncing is paused in Syncer")
					}
					return errors.New("Still syncing; the other PC may be off")
				}
			}
			sc.Progress("Up to date")
			return nil
		}}
}

// savesAfterStep backs the game's saves up once it has exited. With wait
// false it only starts the backup (Syncer finishes it on its own), for a
// game started elsewhere, where nobody is waiting on a launch sequence.
func (c *Core) savesAfterStep(g library.Game, known *bool, wait bool) launch.Step {
	return launch.Step{ID: "savesAfter", Label: "Back up saves", Timeout: 3 * time.Minute,
		Run: func(ctx context.Context, sc *launch.StepContext) error {
			start := c.Settings.Get().StartSyncer
			dctx, cancel := context.WithTimeout(ctx, 15*time.Second)
			cl, err := syncer.Dial(dctx, start)
			cancel()
			if errors.Is(err, syncer.ErrOutdated) {
				return err
			}
			if err != nil && !start {
				sc.Progress("Syncer isn't running")
				return nil
			}
			if err != nil {
				return errors.New("Syncer didn't answer")
			}
			defer cl.Close()
			if !*known {
				// A first play may have made the save folder; Syncer finds
				// new games on its own, so only back up what it knows.
				gs, err := cl.GameStatus(ctx, syncerGame(g))
				if err != nil {
					return err
				}
				if !gs.Known {
					sc.Progress("Syncer has no saves for this game")
					return nil
				}
			}
			sc.Progress("Backing up…")
			if !wait {
				_, err := cl.BackupNow(ctx, false, 0)
				if err != nil && !strings.Contains(strings.ToLower(err.Error()), "already") {
					return err
				}
				sc.Progress("Backing up in the background")
				return nil
			}
			r, err := cl.BackupNow(ctx, true, 2*time.Minute+30*time.Second)
			switch {
			case err != nil && strings.Contains(strings.ToLower(err.Error()), "already"):
				sc.Progress("Syncer is already backing up")
			case err != nil:
				return err
			case !r.Finished:
				sc.Progress("Backing up in the background")
			case !r.OK && len(r.Errors) > 0:
				return errors.New(strings.TrimSpace(r.Errors[0]))
			default:
				sc.Progress("Backed up")
			}
			return nil
		}}
}

File: frontend/src/lib/saves.ts
// A game's saves in words, from what Syncer reports.
import { ago } from "./format";
import type { Saves, SyncerStatus } from "./types";

export interface SyncerSummary {
  /** "Connected", "Not running", … */
  title: string;
  detail: string;
  tone: "ok" | "warn" | "muted";
  /** The one thing to do about it, if anything. */
  action?: "get" | "update" | "start" | "open";
}

/** Syncer itself in words, for Settings. */
export function syncerSummary(s: SyncerStatus | null, startSyncer: boolean): SyncerSummary | null {
  if (!s) return null;
  if (!s.installed)
    return {
      title: "Not installed",
      detail: "Syncer keeps game saves in sync between your PCs and backs them up. With it, WaterLauncher brings a game's saves up to date before you play and backs them up after.",
      tone: "muted",
      action: "get",
    };
  if (s.outdated)
    return {
      title: "Needs an update",
      detail: `Syncer ${s.version ?? ""} is installed; WaterLauncher works with 0.11 or newer.`.replace("  ", " "),
      tone: "warn",
      action: "update",
    };
  if (!s.running)
    return {
      title: "Not running",
      detail: startSyncer
        ? "WaterLauncher starts it in the background when a game needs its saves."
        : "WaterLauncher won't start it, so games start without syncing their saves until you open Syncer.",
      tone: startSyncer ? "muted" : "warn",
      action: "start",
    };
  if (!s.connected) return { title: "Didn't answer", detail: s.error || "Syncer is running but didn't answer. Try again in a moment.", tone: "warn", action: "open" };
  const bits = [`Syncer ${s.version ?? ""}`.trim(), `${s.games} ${s.games === 1 ? "game" : "games"}`];
  if (s.lastBackup) bits.push(`backed up ${ago(s.lastBackup).toLowerCase()}`);
  if (s.conflicts)
    return {
      title: s.conflicts === 1 ? "A save has two versions" : `${s.conflicts} saves have two versions`,
      detail: `${bits.join(" · ")}. Pick the version to keep in Syncer; until then WaterLauncher asks before starting that game.`,
      tone: "warn",
      action: "open",
    };
  if (s.paused)
    return {
      title: s.pausedUntil ? `Syncing paused until ${new Date(s.pausedUntil * 1000).toLocaleTimeString([], { hour: "2-digit", minute: "2-digit" })}` : "Syncing paused",
      detail: `${bits.join(" · ")}. Saves aren't brought up to date before you play while syncing is paused.`,
      tone: "warn",
      action: "open",
    };
  if (!s.syncing)
    return { title: "Connected, not syncing", detail: `${bits.join(" · ")}. Syncer backs saves up, but its syncing isn't running.`, tone: "warn", action: "open" };
  return { title: s.backingUp ? "Connected · backing up" : "Connected", detail: `${bits.join(" · ")}.`, tone: "ok" };
}

export type SavesTone = "ok" | "warn" | "muted";

export interface SavesSummary {
  /** Short line: "Synced · backed up 2 h ago". */
  text: string;
  /** Longer explanation for details views. */
  detail: string;
  tone: SavesTone;
  /** What the button next to it does, if anything. */
  action?: "get" | "open";
}

const unix = (iso: string) => {
  const t = Date.parse(iso);
  return Number.isNaN(t) || t <= 0 ? 0 : Math.floor(t / 1000);
};

function backedUpText(iso: string): string {
  const t = unix(iso);
  return t ? `backed up ${ago(t).toLowerCase()}` : "not backed up yet";
}

export function savesSummary(s: Saves | null): SavesSummary | null {
  if (!s) return null;
  if (!s.installed)
    return {
      text: "Syncer isn't installed",
      detail: "Syncer keeps saves in sync between your PCs and backs them up. WaterLauncher then syncs them before you play.",
      tone: "muted",
      action: "get",
    };
  if (s.outdated)
    return {
      text: "Syncer needs an update",
      detail: "WaterLauncher works with Syncer 0.11 or newer: it then syncs a game's saves before you play and backs them up after.",
      tone: "warn",
      action: "get",
    };
  if (!s.available) return { text: "Syncer didn't answer", detail: s.error || "Try again in a moment.", tone: "warn", action: "open" };
  if (!s.known) return { text: "Not in Syncer", detail: "Syncer has no saves for this game (yet). It picks up new games on its own.", tone: "muted", action: "open" };

  const conflicts = s.folders.reduce((n, f) => n + f.conflicts, 0);
  if (conflicts)
    return {
      text: conflicts === 1 ? "2 versions of a save" : `${conflicts} saves with 2 versions`,
      detail: "Two PCs changed the same save. Pick the version to keep in Syncer before you play.",
      tone: "warn",
      action: "open",
    };
  const newer = s.folders.find((f) => f.newerOn);
  if (newer)
    return {
      text: `Newer save on ${newer.newerOn}`,
      detail: `${newer.newerOn} backed up a newer save that hasn't reached this PC yet. Turn that PC on, or wait for it when you play.`,
      tone: "warn",
      action: "open",
    };

  const synced = s.folders.filter((f) => f.sync);
  const latest = s.folders.map((f) => f.backedUp).sort().at(-1) ?? "";
  if (!synced.length) {
    const b = backedUpText(latest);
    return { text: `${b[0].toUpperCase()}${b.slice(1)} · not synced`, detail: "Syncer backs these saves up but doesn't sync them to other PCs.", tone: "ok" };
  }
  const state = synced.some((f) => f.state === "paused")
    ? "Syncing paused"
    : synced.some((f) => f.state !== "idle" || f.needBytes > 0)
      ? "Syncing…"
      : "Synced";
  return {
    text: `${state} · ${backedUpText(latest)}`,
    detail: "Syncer syncs these saves with your other PCs and backs them up. They're brought up to date before you play and backed up after.",
    tone: state === "Synced" ? "ok" : "warn",
  };
}
